text/0005-is-operator.md

```diff
@@ -0,0 +1,176 @@
+
+# Motivation
+
+The `is` operator in useful in (at least) two common scenarios.
+
+## Use case 1: Input validation
+
+Say that a user wants to allow any `User` to view files in the folder `Folder::"public"`.
+Currently, to achieve the desired behavior, the user should define a schema with the action entity "view" that can apply to principals of type `User` and resources of type `File`, and then write the following policy.
+
+```
+permit(principal, action == Action::"view", resource in Folder::"public");
+```
+
+Assuming that they pass in inputs where the principal is of type `User` and the resource is of type `File`, this will work as expected.
+However, it could also return "allow" for other principal types like `Folder` or resource types like `User`.
+This is a bug on the user's end: they shouldn’t make a query with the action "view" when the principal is not a `User` and the resource is not a `File`.
+They are breaking the contract they stated in their schema.
+Nevertheless, it might be confusing that such a query does not result in a "deny".
+
+With an `is` operator, users could instead write the following.
+
+```
+permit(
+  principal is User,
+  action == Action::"view",
+  resource is File in Folder::"public"
+);
+```
+
+`is` acts as a dynamic check on the runtime types of `principal` and `resource`.
+This puts the important type information into the policy itself, as opposed to a separate schema, making it easy to tell what the policy does.
+
+## Use case 2: Reduce `has` checks
+
+The `is` operator can also be used in some places where `has` is currently required.
+For example, consider the following policy taken from the Cedar [document cloud example](https://github.com/cedar-policy/cedar-examples/tree/main/cedar-example-use-cases/document_cloud).
+
+```
+forbid (principal, action, resource)
+when
+{
+  resource has owner &&
+  principal != resource.owner &&
+  resource has isPrivate &&
+  resource.isPrivate
+};
+```
+
+This policy acts on any type of resource, but can only be satisfied when the resource has `owner` and `isPrivate` attributes.
+If the resource does not have these attributes, then the policy is a no-op.
+Either way, we need to check whether the attributes exist using `has` to avoid runtime errors.
+
+Now say that only one entity type (e.g., `File`) has the fields `owner` and `isPrivate`.
+Then this policy could be written more succinctly as follows.
+
+```
+forbid (principal, action, resource is File)
+when
+{
+  principal != resource.owner &&
+  resource.isPrivate
+};
+```
+
+# Detailed design
+
+## Changes required to parser/AST
+
+- Update the grammar and AST to support expressions of the form `e is et` where `et` is a `Name` (e.g., `Namespace::EntityType`).
+- Update the grammar and AST to support expressions of the form `e1 is et in e2` (but not `e1 in e2 is et`, because this is more difficult to mentally parse).
+- Update the grammar to allow expressions like `e is et` and `e1 is et in e2` in the policy scope.
```

Comment:
Do we need to include `is` in the policy scope, and do we need to include it as part of `in` expression? We don't allow things like `resource has public` in the policy scope, nor do we allow expressions like `resource has public in Folder::"public"`. What necessitates putting `is` there? I would argue that it doesn't make sense to do that because `is` expressions will not be used for policy indexing.

Reply:
> because `is` expressions will not be used for policy indexing.

Why wouldn't they be?  It seems to me like some consumer of Cedar could easily choose to index on `is` expressions.  In fact this might be a good idea from a performance perspective; when you see an authz query for a resource of type `XYZ`, you can use indexing to avoid pulling policies that use `resource is` with any type other than `XYZ`.  Seems to me that we should leave this door easily open to Cedar users.

Reply:
You could do policy indexing on attributes too. Should we allow attributes in the scope?

Reply:
Maybe we should!  Write the RFC :)

Reply:
More seriously, we should have a story for why we include certain expressions and not others that are legal in the policy scope.  And then apply that general principle here.

Reply:
The general principle revolves around indexing based on "edges" between _entities_ in the policy store that are on the RHS of either `==` or `in`, e.g., `User::"Alice"` or `"Album::"Photos"`. Since a type like `Photo` is not an entity, I'm scratching my head a little to see how it would fit in this scheme. Even if it did, what would it mean to allow `resource is Photo in Album::"Photos"`?  You'd potentially have a hyper-edge.

Since we don't have a good reason right now to include `is` in indexing, it seems simpler and more conceptually pure to leave `is` in the `when` clause, at least for now, and then revisit with experience. Two-way door.

Reply:
I concede your point about indexing.

I did some more thinking about why it "feels" to me like we should allow `is` in the policy scope.  I came up with two arguments; maybe neither of them are any good :) and I acknowledge your point that we can always disallow it for now and later decide to add it back-compatibly.  But here goes.

First, I think users will want and expect that they can use `is` in the policy scope.  Features isomorphic to `is` have been requested many times, and in most (all?) of those cases, the motivating example presents `is` in the policy scope.  It seems to be the intuitive place for it, to our users.  If we don't allow it, users will certainly ask for it.

Second, an `is` restriction on `principal` or `resource` in a policy feels like a fundamental part of the policy you'd want up front.  I would argue that the policy scope "feels" in some way different than the `when` clause, in that the scope "defines which principals/resources we're even talking about" while the `when` clause performs some checks on those principals/resources.  Using `is` to restrict a policy to only apply to `User` principals or `File` resources feels like a fundamental part of the policy, feels like defining which principals/resources we're even talking about, rather than feeling like a check being performed on a principal/resource we've already selected.  To me, at least.  And I think putting those things up top makes intuitive sense, much like how the type signature of a function is put on top, before its body.

Again I recognize that neither of these arguments are very strong :) The argument that we shouldn't allow `is` in the policy scope certainly has more principled support, namely that it can be grounded in an argument about indexing, and it's the two-way-door option.  But just putting my brainstorming on paper here.

Reply:
I also think `is` makes sense in the policy scope (for readability, at least). Similar to `in` expressions, `is` expressions restrict the "scope" of what the policy applies to, and set the context for any other restrictions in the policy condition. I feel like `is` is different from `has` since it directly references an entity type defined in the schema, which will restrict the types of entities that it can be related to by membership.

Re indexing: I think I need more context/definitions here to understand the problem. Do we have an example written up somewhere?

----

I appreciate the point that deciding not to allow `is` in the policy scope is something that we can easily reverse later, but I think it's best to discuss/decide now rather than kicking the can down the road. Like @cdisselkoen said, every time a user has asked for this, their example policy used `is` in the scope, so we should expect this request.

Reply:
> it's best to discuss/decide now rather than kicking the can down the road.

I'm not sure I agree. Amazon culture favors quick decisionmaking for a reason. We have limited information now, so more talking doesn't necessarily help. Making a two-way door decision allows customers to start using what we add, and from that usage we can get more feedback on the parts we don't.

Reply:
One potentially irrelevant nitpick is that the fact that we can't index on type membership seems to me like it's more of a quirk of dynamodb, our data model, our indexing strategy, and our security practices (hashing everything to protect customer data from being visible) than it is a limitation of indexing itself. I could be wrong about this but i believe one could design an indexing strategy similar to R-trees that expressed entity type membership just like R trees express containments of points in space. 

It may not work in practice at all because EntityUIDs might not be subdivisible enough, I don't know. If that didn't work my next idea would be a reverse index akin to those used in full text search (also not supported by ddb). But there may be some other ways to cleverly index them that I haven't thought about. Point is that even if AVP can't do something right now because of our db schema, this doesn't mean we won't ever be able to do it, and it also doesn't mean customers can't do it.